File: src/backend/core/data_manager.py
import os
import json
import shutil
import tempfile
import threading
import collections.abc
from collections import defaultdict
from pathlib import Path

_file_locks = defaultdict(threading.Lock)

def atomic_write_json(data, target_file):
    target_file = Path(target_file)
    fd, temp_path = tempfile.mkstemp(dir=target_file.parent, prefix="._tmp_", suffix=".json")
    try:
        with os.fdopen(fd, "w", encoding="utf-8") as f:
            json.dump(data, f, ensure_ascii=False, indent=2)
        os.replace(temp_path, target_file)
    except Exception as e:
        if os.path.exists(temp_path):
            os.remove(temp_path)
        raise e
BASE_DIR = Path(__file__).parent.parent
DB_DIR = BASE_DIR / "saved_projects"

def get_project_dir(project_id: str) -> Path:
    return DB_DIR / project_id

def init_project(project_id: str, blueprint_data: dict, modules_dir_path: str, full_comp_desc: dict):
    p_dir = get_project_dir(project_id)
    p_dir.mkdir(parents=True, exist_ok=True)
    m_dir = p_dir / "modules"
    m_dir.mkdir(exist_ok=True)
    
    # Save Blueprint
    with open(p_dir / "blueprint_CompDesc.json", "w", encoding="utf-8") as f:
        json.dump(blueprint_data, f, ensure_ascii=False, indent=2)
    
    # Save Full Enriched JSON (for CSV generation and reference)
    with open(p_dir / "CompDesc.json", "w", encoding="utf-8") as f:
        json.dump(full_comp_desc, f, ensure_ascii=False, indent=2)

    src_modules = Path(modules_dir_path)
    if src_modules.exists():
        for item in src_modules.iterdir():
            if item.is_file() and item.name.endswith(".json"):
                shutil.copy2(item, m_dir / item.name)
    


def deep_update(d, u, path="root"):
    """Robust deep update with dual key (Snake/Camel) support for COMBOX"""
    for k, v in u.items():
        curr_path = f"{path}.{k}"
        
        # 1. Handle List merging (AttributeGroups, elements, interface_group)
        if k in d and isinstance(d[k], list) and isinstance(v, list):
            d_list = d[k]
            for u_item in v:
                if isinstance(u_item, dict):
                    # Keyed search: supports key, type, interfaceUuid, stringValue etc.
                    ukey = u_item.get("key") or u_item.get("type") or u_item.get("interfaceUuid") or u_item.get("interface_uuid")
                    if ukey:
                        match = next((item for item in d_list if isinstance(item, dict) and (item.get("key") == ukey or item.get("type") == ukey or item.get("interfaceUuid") == ukey or item.get("interface_uuid") == ukey)), None)
                        if match:
                            deep_update(match, u_item, f"{curr_path}[key:{ukey}]")
                        else:
                            d_list.append(u_item)
                    else:
                        d_list.append(u_item)
        
        # 2. Handle Dictionary (Branch) merging
        elif isinstance(v, collections.abc.Mapping):
            # ━━━ CRITICAL FIX: DUAL KEY COMBOX PROTECTION ━━━
            is_combox = k in ["comboType", "combo_type"]
            if is_combox:
                t_key = v.get("typeKey") or v.get("type_key")
                if t_key and ("typeGroups" not in v and "type_groups" not in v):
                    # Only updating selection
                    if k in d and isinstance(d[k], dict):
                        target_key = "typeKey" if "typeKey" in d[k] else "type_key"
                        if d[k].get(target_key) != t_key:
                            print(f"DISK_AUDIT: [COMBOX_CHANGE] {curr_path}.{target_key}: [{d[k].get(target_key)}] -> [{t_key}]", flush=True)
                            d[k][target_key] = t_key
                        continue
            
            # Normal recursion
            deep_update(d.get(k, {}), v, curr_path)
            
        # 3. Handle Leaf values
        else:
            if d.get(k) != v:
                print(f"DISK_AUDIT: [VALUE_CHANGE] {curr_path}: [{d.get(k)}] -> [{v}]", flush=True)
                d[k] = v
    return d

def ensure_module_in_project(project_id: str, module_filename: str, fallback_source_path: Path) -> bool:
    """Ensures a module file exists in the project sandbox."""
    p_dir = get_project_dir(project_id)
    m_dir = p_dir / "modules"
    os.makedirs(str(m_dir), exist_ok=True)
    target = m_dir / module_filename
    if not target.exists():
        if fallback_source_path.exists():
            shutil.copy2(fallback_source_path, target)
            print(f"DISK_AUDIT: [SANDBOX_IMPORT] Copied {module_filename} to project {project_id}", flush=True)
            return True
    return target.exists()

def update_component(project_id: str, module_uuid: str, payload_delta: dict) -> bool:
    m_dir = get_project_dir(project_id) / "modules"
    target_file = next(m_dir.glob(f"*{module_uuid}*.json"), None)
    if not target_file: return False
    with _file_locks[str(target_file)]:
        with open(target_file, "r", encoding="utf-8") as file:
            data = json.load(file)
        print(f"DISK_AUDIT: >>> Updating component {module_uuid} <<<", flush=True)
        deep_update(data, payload_delta)
        atomic_write_json(data, target_file)
    return True

def update_ability(project_id: str, payload_delta: dict) -> bool:
    fpath = get_project_dir(project_id) / "AbiSet.json"
    if not fpath.exists():
        # [O-7] Create from baseline if missing during update
        baseline = BASE_DIR / "resources" / "AbiSet_base.json"
        if baseline.exists():
            shutil.copy2(baseline, fpath)
        else:
            with open(fpath, "w", encoding="utf-8") as f: json.dump({"version": "V1.0"}, f)
            
    with _file_locks[str(fpath)]:
        with open(fpath, "r", encoding="utf-8") as file:
            data = json.load(file)
        print(f"DISK_AUDIT: >>> Updating AbilitySet <<<", flush=True)
        deep_update(data, payload_delta)
        atomic_write_json(data, fpath)
    return True

def update_function(project_id: str, payload_delta: dict) -> bool:
    """[O-9] Manage FuncDesc.json independently."""
    fpath = get_project_dir(project_id) / "FuncDesc.json"
    if not fpath.exists():
        with open(fpath, "w", encoding="utf-8") as f: json.dump({"version": "V1.0", "function": []}, f)
            
    with _file_locks[str(fpath)]:
        with open(fpath, "r", encoding="utf-8") as file:
            data = json.load(file)
        print(f"DISK_AUDIT: >>> Updating FuncDesc <<<", flush=True)
        deep_update(data, payload_delta)
        atomic_write_json(data, fpath)
    return True

def get_component(project_id: str, module_uuid: str):
    m_dir = get_project_dir(project_id) / "modules"
    for f in m_dir.glob(f"*{module_uuid}*.json"):
        with open(f, "r", encoding="utf-8") as file: return json.load(file)
    return None

def get_ability(project_id: str):
    fpath = get_project_dir(project_id) / "AbiSet.json"
    if fpath.exists():
        with open(fpath, "r", encoding="utf-8") as file: return json.load(file)
    return None

def get_function(project_id: str):
    """[O-9] Retrieve FuncDesc.json."""
    fpath = get_project_dir(project_id) / "FuncDesc.json"
    if fpath.exists():
        with open(fpath, "r", encoding="utf-8") as file: return json.load(file)
    return None

File: src/backend_ts/src/core/dataManager.ts
/**
 * core/dataManager.ts
 * ====================
 * AMR Studio V4 - 项目沙箱与数据管理器
 * 该模块负责管理用户在前端操作时产生的局部沙箱项目（Saved Projects）。
 */

import * as fs from 'fs/promises';
import { existsSync } from 'fs';
import * as path from 'path';
import { RESOURCE_DIR, SAVED_PROJECTS_DIR } from './paths';

// 简单的高并发异步锁，用于确保 Node.js 单进程环境下的文件写入顺序安全性
class FileLock {
  private static locks = new Map<string, Promise<void>>();

  static async acquire(filePath: string): Promise<() => void> {
    const previous = this.locks.get(filePath) || Promise.resolve();
    let resolveLock!: () => void;
    const next = new Promise<void>((resolve) => {
      resolveLock = resolve;
    });
    this.locks.set(filePath, next);
    await previous;
    return () => {
      resolveLock();
      if (this.locks.get(filePath) === next) {
        this.locks.delete(filePath);
      }
    };
  }
}

/**
 * 原子级 JSON 文件写入器：
 * 首先写入临时文件中，并在成功写入后通过系统重命名操作原子替换原文件。
 */
export async function atomicWriteJson(data: any, targetFile: string): Promise<void> {
  const fileDir = path.dirname(targetFile);
  const tempFile = path.join(fileDir, `._tmp_${Date.now()}_${Math.random().toString(36).substring(2, 7)}.json`);
  
  try {
    await fs.mkdir(fileDir, { recursive: true });
    await fs.writeFile(tempFile, JSON.stringify(data, null, 2), 'utf-8');
    await fs.rename(tempFile, targetFile);
  } catch (err) {
    if (existsSync(tempFile)) {
      await fs.unlink(tempFile).catch(() => {});
    }
    throw err;
  }
}

export const DB_DIR = SAVED_PROJECTS_DIR;

export function getProjectDir(projectId: string): string {
  return path.join(DB_DIR, projectId);
}

/**
 * 项目初始化主逻辑
 */
export async function initProject(projectId: string, blueprintData: any, modulesDirPath: string, fullCompDesc: any): Promise<void> {
  const pDir = getProjectDir(projectId);
  const mDir = path.join(pDir, 'modules');
  
  await fs.mkdir(pDir, { recursive: true });
  await fs.mkdir(mDir, { recursive: true });

  // 1. 保存骨架蓝图文件
  await fs.writeFile(path.join(pDir, 'blueprint_CompDesc.json'), JSON.stringify(blueprintData, null, 2), 'utf-8');

  // 2. 保存合并的完整数据，作为基准参照
  await fs.writeFile(path.join(pDir, 'CompDesc.json'), JSON.stringify(fullCompDesc, null, 2), 'utf-8');

  // 3. 将各拆分出来的组件 JSON 配置文件拷贝到该项目沙箱的 modules 物理目录中
  if (existsSync(modulesDirPath)) {
    const items = await fs.readdir(modulesDirPath);
    for (const item of items) {
      const srcFile = path.join(modulesDirPath, item);
      const stat = await fs.stat(srcFile);
      if (stat.isFile() && item.endsWith('.json')) {
        await fs.copyFile(srcFile, path.join(mDir, item));
      }
    }
  }
}

/**
 * 高精度深度合并算法
 */
export function deepUpdate(d: any, u: any, pathStr: string = 'root'): any {
  if (typeof u !== 'object' || u === null) return d;

  for (const [k, v] of Object.entries(u)) {
    const currPath = `${pathStr}.${k}`;

    // 1. 列表合并处理（例如属性组列表 AttributeGroups、子组件属性列表、接口组 interfaceGroup 等）
    if (k in d && Array.isArray(d[k]) && Array.isArray(v)) {
      const dList = d[k];
      for (const uItem of v) {
        if (typeof uItem === 'object' && uItem !== null) {
          const ukey = uItem.key || uItem.type || uItem.interfaceUuid || uItem.interface_uuid;
          if (ukey) {
            const matchIndex = dList.findIndex((item: any) => {
              return typeof item === 'object' && item !== null &&
                (item.key === ukey || item.type === ukey || item.interfaceUuid === ukey || item.interface_uuid === ukey);
            });
            if (matchIndex !== -1) {
              deepUpdate(dList[matchIndex], uItem, `${currPath}[key:${ukey}]`);
            } else {
              dList.push(uItem);
            }
          } else {
            dList.push(uItem);
          }
        } else {
          dList.push(uItem);
        }
      }
    }
    // 2. 字典分支合并处理
    else if (typeof v === 'object' && v !== null && !Array.isArray(v)) {
      // ━━━ 特殊安全防护：COMBOX 双键选择项修改防护 ━━━
      const isCombox = k === 'comboType' || k === 'combo_type';
      if (isCombox) {
        const tKey = (v as any).typeKey || (v as any).type_key;
        if (tKey && !('typeGroups' in v) && !('type_groups' in v)) {
          if (k in d && typeof d[k] === 'object' && d[k] !== null) {
            const targetKey = 'typeKey' in d[k] ? 'typeKey' : 'type_key';
            if (d[k][targetKey] !== tKey) {
              console.log(`DISK_AUDIT: [COMBOX_CHANGE] ${currPath}.${targetKey}: [${d[k][targetKey]}] -> [${tKey}]`);
              d[k][targetKey] = tKey;
            }
            continue;
          }
        }
      }

      // 正常分支递归深度合并
      let existing = d[k];
      if (typeof existing !== 'object' || existing === null || Array.isArray(existing)) {
        existing = {};
        d[k] = existing;
      }
      deepUpdate(existing, v, currPath);
    }
    // 3. 叶子单元值更新处理
    else {
      if (d[k] !== v) {
        console.log(`DISK_AUDIT: [VALUE_CHANGE] ${currPath}: [${d[k]}] -> [${v}]`);
        d[k] = v;
      }
    }
  }
  return d;
}

/**
 * 确保沙箱内的组件就绪
 */
export async function ensureModuleInProject(projectId: string, moduleFilename: string, fallbackSourcePath: string): Promise<boolean> {
  const pDir = getProjectDir(projectId);
  const mDir = path.join(pDir, 'modules');
  await fs.mkdir(mDir, { recursive: true });
  
  const target = path.join(mDir, moduleFilename);
  if (!existsSync(target)) {
    if (existsSync(fallbackSourcePath)) {
      await fs.copyFile(fallbackSourcePath, target);
      console.log(`DISK_AUDIT: [SANDBOX_IMPORT] Copied ${moduleFilename} to project ${projectId}`);
      return true;
    }
  }
  return existsSync(target);
}

/**
 * 线程（并发）安全地更新沙箱中的组件
 */
export async function updateComponent(projectId: string, moduleUuid: string, payloadDelta: any): Promise<boolean> {
  const mDir = path.join(getProjectDir(projectId), 'modules');
  if (!existsSync(mDir)) return false;

  const files = await fs.readdir(mDir);
  const targetFile = files.find(f => f.includes(moduleUuid) && f.endsWith('.json'));
  if (!targetFile) return false;

  const targetPath = path.join(mDir, targetFile);
  const release = await FileLock.acquire(targetPath);
  try {
    const raw = await fs.readFile(targetPath, 'utf-8');
    const data = JSON.parse(raw);
    console.log(`DISK_AUDIT: >>> Updating component ${moduleUuid} <<<`);
    deepUpdate(data, payloadDelta);
    await atomicWriteJson(data, targetPath);
    return true;
  } finally {
    release();
  }
}

/**
 * 并发安全地更新沙箱的能力配置
 */
export async function updateAbility(projectId: string, payloadDelta: any): Promise<boolean> {
  const fpath = path.join(getProjectDir(projectId), 'AbiSet.json');
  
  if (!existsSync(fpath)) {
    const baseline = path.join(RESOURCE_DIR, 'AbiSet_base.json');
    if (existsSync(baseline)) {
      await fs.copyFile(baseline, fpath);
    } else {
      await fs.writeFile(fpath, JSON.stringify({ version: 'V1.0' }, null, 2));
    }
  }

  const release = await FileLock.acquire(fpath);
  try {
    const raw = await fs.readFile(fpath, 'utf-8');
    const data = JSON.parse(raw);
    console.log(`DISK_AUDIT: >>> Updating AbilitySet <<<`);
    deepUpdate(data, payloadDelta);
    await atomicWriteJson(data, fpath);
    return true;
  } finally {
    release();
  }
}

/**
 * 并发安全地更新沙箱的动作逻辑描述
 */
export async function updateFunction(projectId: string, payloadDelta: any): Promise<boolean> {
  const fpath = path.join(getProjectDir(projectId), 'FuncDesc.json');

  if (!existsSync(fpath)) {
    await fs.writeFile(fpath, JSON.stringify({ version: 'V1.0', function: [] }, null, 2));
  }

  const release = await FileLock.acquire(fpath);
  try {
    const raw = await fs.readFile(fpath, 'utf-8');
    const data = JSON.parse(raw);
    console.log(`DISK_AUDIT: >>> Updating FuncDesc <<<`);
    deepUpdate(data, payloadDelta);
    await atomicWriteJson(data, fpath);
    return true;
  } finally {
    release();
  }
}

/**
 * 获取组件 JSON
 */
export async function getComponent(projectId: string, moduleUuid: string): Promise<any | null> {
  const mDir = path.join(getProjectDir(projectId), 'modules');
  if (!existsSync(mDir)) return null;

  const files = await fs.readdir(mDir);
  const targetFile = files.find(f => f.includes(moduleUuid) && f.endsWith('.json'));
  if (!targetFile) return null;

  const raw = await fs.readFile(path.join(mDir, targetFile), 'utf-8');
  return JSON.parse(raw);
}

/**
 * 获取能力配置 JSON
 */
export async function getAbility(projectId: string): Promise<any | null> {
  const fpath = path.join(getProjectDir(projectId), 'AbiSet.json');
  if (existsSync(fpath)) {
    const raw = await fs.readFile(fpath, 'utf-8');
    return JSON.parse(raw);
  }
  return null;
}

/**
 * 获取动作功能配置 JSON
 */
export async function getFunction(projectId: string): Promise<any | null> {
  const fpath = path.join(getProjectDir(projectId), 'FuncDesc.json');
  if (existsSync(fpath)) {
    const raw = await fs.readFile(fpath, 'utf-8');
    return JSON.parse(raw);
  }
  return null;
}
